From 54fc4977b608b8ab5df428ea319a27170e29fb25 Mon Sep 17 00:00:00 2001
From: Stefan Assmann <sassmann@redhat.com>
Date: Tue, 2 Sep 2014 11:39:45 -0400
Subject: [ethernet] i40e: FD filter replay logic bug fix

Message-id: <1409658042-22689-225-git-send-email-sassmann@redhat.com>
Patchwork-id: 90425
O-Subject: [RHEL7.1 PATCH 224/281] i40e: FD filter replay logic bug fix
Bugzilla: 1091126
RH-Acked-by: Tony Camuso <tcamuso@redhat.com>

From: Anjali Singhai Jain <anjali.singhai@intel.com>

With the auto_disable flags added there was a bug that was causing the
replay logic to not work correctly.
This patch fixes the issue so that we call a replay after a sideband
reset correctly.

Change-ID: I005fe1ac361188ee5b19517a83c922038cba1b00
Signed-off-by: Anjali Singhai Jain <anjali.singhai@intel.com>
Signed-off-by: Jeff Kirsher <jeffrey.t.kirsher@intel.com>
Signed-off-by: Stefan Assmann <sassmann@redhat.com>
(cherry picked from commit b814ba65fc625c6791987329737a4dc5b6c95566)
---

Signed-off-by: Jarod Wilson <jarod@redhat.com>

diff --git a/drivers/net/ethernet/intel/i40e/i40e_txrx.c b/drivers/net/ethernet/intel/i40e/i40e_txrx.c
index 5cc27fb..babb7e6 100644
--- a/drivers/net/ethernet/intel/i40e/i40e_txrx.c
+++ b/drivers/net/ethernet/intel/i40e/i40e_txrx.c
@@ -445,14 +445,16 @@ static void i40e_fd_handle_status(struct i40e_ring *rx_ring,
    */
   if (fcnt_prog >= (fcnt_avail - I40E_FDIR_BUFFER_FULL_MARGIN)) {
    /* Turn off ATR first */
-   if (pf->flags & I40E_FLAG_FD_ATR_ENABLED) {
-    pf->flags &= ~I40E_FLAG_FD_ATR_ENABLED;
+   if ((pf->flags & I40E_FLAG_FD_ATR_ENABLED) &&
+       !(pf->auto_disable_flags &
+         I40E_FLAG_FD_ATR_ENABLED)) {
     dev_warn(&pdev->dev, "FD filter space full, ATR for further flows will be turned off\n");
     pf->auto_disable_flags |=
              I40E_FLAG_FD_ATR_ENABLED;
     pf->flags |= I40E_FLAG_FDIR_REQUIRES_REINIT;
-   } else if (pf->flags & I40E_FLAG_FD_SB_ENABLED) {
-    pf->flags &= ~I40E_FLAG_FD_SB_ENABLED;
+   } else if ((pf->flags & I40E_FLAG_FD_SB_ENABLED) &&
+       !(pf->auto_disable_flags &
+         I40E_FLAG_FD_SB_ENABLED)) {
     dev_warn(&pdev->dev, "FD filter space full, new ntuple rules will not be added\n");
     pf->auto_disable_flags |=
        I40E_FLAG_FD_SB_ENABLED;
-- 
1.7.1